pallets/permissions/src/lib.rs

```diff
@@ -0,0 +1,281 @@
+// Copyright 2021 Centrifuge Foundation (centrifuge.io).
+//
+// This file is part of the Centrifuge chain project.
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate frame_system;
+
+///! A crate that defines a simple permissions logic for our infrastructure.
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+use frame_support::traits::Contains;
+use frame_support::{dispatch::DispatchResult, pallet_prelude::*};
+use frame_system::pallet_prelude::*;
+
+pub trait Permissions<AccountId> {
+	type Location;
+	type Role;
+	type Error;
+
+	fn has_permission(location: Self::Location, who: AccountId, role: Self::Role) -> bool;
+
+	fn add_permission(
+		editor: AccountId,
+		location: Self::Location,
+		who: AccountId,
+		role: Self::Role,
+	) -> Result<(), Self::Error>;
+
+	fn rm_permission(
+		editor: AccountId,
+		location: Self::Location,
+		who: AccountId,
+		role: Self::Role,
+	) -> Result<(), Self::Error>;
+}
+
+pub trait Properties {
+	type Property;
+
+	fn exists(&self, property: Self::Property) -> bool;
+
+	fn empty(&self) -> bool;
+
+	fn rm(&mut self, property: Self::Property);
+
+	fn add(&mut self, property: Self::Property);
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	/// Configure the pallet by specifying the parameters and types on which it depends.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		type Location: Member + Parameter;
+
+		type Role: Member + Parameter;
+
+		type Storage: Member + Parameter + Properties<Property = Self::Role> + Default;
+
+		type Editors: Contains<(Self::AccountId, Self::Role)>;
```

Comment:
Not sure I follow. Do you propose a new Trait? I would not use the interface of `pallet-permissions` for different use-cases. I think this one should be pretty general and only be used by admins. 

As you suggested, you wanted to be safe, that only a specific pallet can change the permissions. This will be checked now in the `Permissions` trait.

Reply:
>  I would not use the interface of pallet-permissions for different use-cases. I think this one should be pretty general and only be used by admins.

I do not follow this. Editors are admins
For example as mentioned here - https://centrifuge.hackmd.io/Lga8cxI9Tia-u0ZPXrt84g#Notes
A poolAdmin and MemberlistAdmin can add new tranche investors for a given pool.
In that case, we want to ensure, they alone can edit them either through permissions extrinsics or through permissions trait used by other pallets. So that other pallets use pemission traits mostly yo check permissions rather than add new ones

Reply:
Oh, this was misleading. With admin I meant the `AdminOrigin` of the pallet, which will be root, democracy or council. The Editors are pallets or whatever we put into Contians call. 

Hance, the pool-pallet can edit pool roles besides `TranchInvestors` and the pallet-restricted-tokens can edit only `TranchInvestors`